### ISSUE TITLE
Not working with factory ?

### ISSUE THREAD

**Issue by eddycharly** [auxmoney/OpentracingBundle-Guzzle#15]
We have a factory that creates a guzzle client, this factory returns the guzzle client with a handler created given some arguments.

The service looks like this:
```yaml

    admin.microservice.client:
        class: GuzzleHttp\Client
        factory: [\Agriconomie\Microservice\ClientFactory, create]
        arguments:
            $debug: '%kernel.debug%'
            $logger: '@monolog.logger.internal_api'
```

And the factory:
```php
    /**
     * @param bool $debug
     * @param LoggerInterface|null $logger
     */
    public static function create(bool $debug, ?LoggerInterface $logger) {
        $handlerStack = HandlerStack::create();
        $client = new \GuzzleHttp\Client([
            'handler' => $handlerStack
        ]);
        if (null !== $logger) {
            $handlerStack->push(Middleware::log($logger, new MessageFormatter(MessageFormatter::SHORT), LogLevel::INFO));
            if ($debug) {
                $handlerStack->push(Middleware::log($logger, new MessageFormatter(MessageFormatter::DEBUG), LogLevel::DEBUG));
            }
        }
        return $client;
    }
```

I integrated the opentracing guzzle bundle, it breaks here i think:
https://github.com/auxmoney/OpentracingBundle-Guzzle/blob/ced155c3335adf4341e22365470f5aa302dc16ef/DependencyInjection/GuzzleHandlerStackCompilerPass.php#L24-L30

probably because of this:
https://github.com/auxmoney/OpentracingBundle-Guzzle/blob/ced155c3335adf4341e22365470f5aa302dc16ef/DependencyInjection/GuzzleHandlerStackCompilerPass.php#L58-L70

Shouldn't the `process` method take care of the factory use case ?

Thanks.

**Comment by cawolf** [comment]
Yes, you are right. As we do not use factories that often in our codebase, we just missed that use case. Do you want to look into it and provide a PR?

**Comment by eddycharly** [comment]
Sure, i or a colleague will try to open a PR today.

Do you have a recommandation on how implementing it ?
Shall we check for the presence of a factory in the definition and try to wrap the factory call ?

**Comment by cawolf** [comment]
Yes, this would be my go-to approach. If possible, try to decorate it in the container (with `decorates`), so that the original service definition still exists.

**Comment by VincentLemaireAgriconomie** [comment]
i'm really not sure about
```
 $handler->push(new Reference(GuzzleTracingHeaderInjection::class));
```
in our factory we instantiate the handle in a method call so i have to wait to have the client.
if the decorator seems ok to you, the next step for me will be to create replace the services.

thanks for any feedback

**Comment by cawolf** [comment]
Ok, I see where this is heading. I forgot about this important bit:
> "handler" is a constructor only option that cannot be overridden in per/request options.

So if you have a factory like above, I cannot think of a way to intercept the handler stack, which is created inline with a static call.

Wouldn't creating a factory for creating the `HandlerStack` as a service be a workaround? This way, we could create a `HandlerStackFactoryDecorator` and just adding a `push()` to the created `HandlerStack`?

**Comment by eddycharly** [comment]
Isn't it possible to register a decorator ?

The decorator would have to return a `GuzzleHttp\Client` though. If the decorator can be instantiated through a factory it should be easy to plug in:
- let the original service as it is
- register a decorator that takes the original service in input (the `GuzzleHttp\Client`)
- the decorator retrieves the handler from the client as in the @VincentLemaireAgriconomie PR and pushes the opentracing middlewares
- the decorator returns the `GuzzleHttp\Client` that was modified

Could something like this work ?

**Comment by cawolf** [comment]
Yes, it would work. The comment in the constructor threw me off, I didn't think about the opportunity to change the handler via `getConfig()`. So yes, this approach should work, and I left some comments in the draft PR.

**Comment by cawolf** [comment]
🎉 This PR is included in version 0.6.0 🎉
